concordium-consensus/src/Concordium/GlobalState/Persistent/Trie.hs

```diff
@@ -111,60 +188,60 @@ instance Bifunctor (TrieF k) where
 instance (Serialize r, Serialize (Nullable r), Serialize v) => Serialize (TrieF k v r) where
     put (Branch vec) = do
         putWord8 1
-        forM_ vec put
+        forM_ (branchesToList vec) put
     put (Tip v) = putWord8 2 >> put v
     put (Stem l r) = do
-        let len = length l
+        let len = SBS.length l
         if len <= 251 then
             putWord8 (3 + fromIntegral len)
         else do
             putWord8 255
             putWord64be (fromIntegral len)
-        forM_ l putWord8
+        putShortByteString l
         put r
     get = getWord8 >>= \case
         0 -> fail "Empty trie"
-        1 -> Branch . V.fromList <$> replicateM 256 get
+        1 -> Branch . branchesFromList <$> replicateM 256 get
         2 -> Tip <$> get
         v -> do
             len <- if v == 255 then
                     fromIntegral <$> getWord64be
                 else
                     return (fromIntegral (v - 3))
-            Stem <$> replicateM len getWord8 <*> get
+            Stem <$> getShortByteString len <*> get
 
 instance (BlobStorable m r, BlobStorable m (Nullable r), BlobStorable m v) => BlobStorable m (TrieF k v r) where
     storeUpdate (Branch vec) = do
-        pvec <- mapM storeUpdate vec
-        return $!! (putWord8 1 >> sequence_ (fst <$> pvec), Branch (snd <$> pvec))
+        pvec <- mapM storeUpdate (branchesToList vec)
+        return $!! (putWord8 1 >> sequence_ (fst <$> pvec), Branch (branchesFromList $ snd <$> pvec))
```

Comment:
What is the rationale for keeping full branches when store then on disk?

If most branches are not near full then it would save space to store a list of (k, v) pairs. And we'd only need one byte for the length.

Reply:
We could maintain that by making use of the fact that keys in the tries we use are never more than 96 bytes. So we could use tags, e.g. 200 for "branch in k, v representation".

We can do this separately, to test if it saves on disk space. But please provide a comment now as to why this disk layout is chosen for the trie.

Reply:
Noted.